libraries/AC_Autorotation/AC_Autorotation.cpp

```diff
@@ -384,3 +558,160 @@ float AC_Autorotation::calc_speed_forward(void)
     return speed_forward;
 }
 
+void AC_Autorotation::flare_controller()
+{
+
+    // Specify forward velocity component and determine delta velocity with respect to time
+    _speed_forward = calc_speed_forward(); //(cm/s)
+    _delta_speed_fwd = _speed_forward - _speed_forward_last; //(cm/s)
+    _speed_forward_last = _speed_forward; //(cm/s)
+    _desired_speed = linear_interpolate(0.0f, _flare_entry_speed, _est_alt, _cushion_alt, _flare_alt_calc);
+
+    // get p
+    _vel_p = _p_fw_vel.get_p(_desired_speed - _speed_forward);
+
+    //calculate acceleration target based on PI controller
+    _accel_target = _vel_p ;
+
+    // filter correction acceleration
+    _accel_target_filter.set_cutoff_frequency(10.0f);
+    _accel_target_filter.apply(_accel_target, _dt);
+
+    //Limits the maximum change in pitch attitude based on acceleration
+    if (_accel_target > _accel_out_last + _accel_max) {
+        _accel_target = _accel_out_last + _accel_max;
+    } else if (_accel_target < _accel_out_last - _accel_max) {
+        _accel_target = _accel_out_last - _accel_max;
+    }
+
+    //Limiting acceleration based on velocity gained during the previous time step
+    if (fabsf(_delta_speed_fwd) > _accel_max * _dt) {
+        _flag_limit_accel = true;
+    } else {
+        _flag_limit_accel = false;
+    }
+
+    if ((_flag_limit_accel && fabsf(_accel_target) < fabsf(_accel_out_last)) || !_flag_limit_accel) {
+        _accel_out = _accel_target;
+    } else {
+        _accel_out = _accel_out_last;
+    }
+    _accel_out_last = _accel_out;
+
+
+
+    //estimate flare effectiveness
+
+    if (_speed_forward <= (0.6*_flare_entry_speed) && fabsf(_avg_sink_deriv)<=0.005f && _avg_acc_z>= -(1.1*9.80665f) ) {
```

Comment:
```suggestion
    if (_speed_forward <= (0.6*_flare_entry_speed) && fabsf(_avg_sink_deriv)<=0.005f && _avg_acc_z>= -(1.1*GRAVITY_MSS) ) {
```

Reply:
May need an include from AP_Math/definitions.h

Reply:
I would prefer it if we used the already existing landing detector and used land_complete as oppose to implementing a new set of thresholds.

Reply:
> I would prefer it if we used the already existing landing detector and used land_complete as oppose to implementing a new set of thresholds.

the only reason I thought this way is because the heli is still flying at this timestamp. This part of the code should only check if the rotor system is capable of providing any further decrease of rate of descent through "bleeding" forward speed.
So if the rotor has exhausted the braking capability, code can move on at leveling the nose down and getting ready for the final touchdown collective pull.